File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation2/observables/location/RemoveLocationIntentUpdatesObservableOnSubscribe.java
package pl.charmas.android.reactivelocation2.observables.location;

import android.app.PendingIntent;
import androidx.annotation.NonNull;

import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.ResultCallback;
import com.google.android.gms.common.api.Status;
import com.google.android.gms.location.LocationServices;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import pl.charmas.android.reactivelocation2.observables.BaseLocationObservableOnSubscribe;
import pl.charmas.android.reactivelocation2.observables.ObservableContext;
import pl.charmas.android.reactivelocation2.observables.ObservableFactory;
import pl.charmas.android.reactivelocation2.observables.StatusException;


public class RemoveLocationIntentUpdatesObservableOnSubscribe extends BaseLocationObservableOnSubscribe<Status> {
    private final PendingIntent intent;

    public static Observable<Status> createObservable(ObservableContext ctx, ObservableFactory factory, PendingIntent intent) {
        return factory.createObservable(new RemoveLocationIntentUpdatesObservableOnSubscribe(ctx, intent));
    }

    private RemoveLocationIntentUpdatesObservableOnSubscribe(ObservableContext ctx, PendingIntent intent) {
        super(ctx);
        this.intent = intent;
    }

    @Override
    protected void onGoogleApiClientReady(GoogleApiClient apiClient, final ObservableEmitter<? super Status> emitter) {
        LocationServices.FusedLocationApi.removeLocationUpdates(apiClient, intent)
                .setResultCallback(new ResultCallback<Status>() {
                    @Override
                    public void onResult(@NonNull Status status) {
                        if (emitter.isDisposed()) return;
                        if (status.isSuccess()) {
                            emitter.onNext(status);
                            emitter.onComplete();
                        } else {
                            emitter.onError(new StatusException(status));
                        }
                    }
                });
    }
}

File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation2/observables/location/RemoveLocationIntentUpdatesObservableOnSubscribe.kt
package pl.charmas.android.reactivelocation2.observables.location

import android.app.PendingIntent
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.location.FusedLocationProviderClient
import io.reactivex.Maybe
import io.reactivex.MaybeEmitter
import pl.charmas.android.reactivelocation2.observables.BaseLocationMaybeOnSubscribe
import pl.charmas.android.reactivelocation2.observables.MaybeContext
import pl.charmas.android.reactivelocation2.observables.MaybeFactory

class RemoveLocationIntentUpdatesObservableOnSubscribe private constructor(
    private val fusedLocationProviderClient: FusedLocationProviderClient,
    ctx: MaybeContext,
    private val intent: PendingIntent
) : BaseLocationMaybeOnSubscribe<Boolean>(ctx) {
    override fun onGoogleApiClientReady(
        apiClient: GoogleApiClient,
        emitter: MaybeEmitter<in Boolean>
    ) {
        fusedLocationProviderClient.removeLocationUpdates(intent)
            .addOnCompleteListener {
                if (!emitter.isDisposed) {
                    emitter.onSuccess(true)
                }
            }
            .addOnFailureListener { error ->
                if (!emitter.isDisposed) {
                    emitter.onError(error)
                }
            }
    }

    companion object {
        fun create(
            fusedLocationProviderClient: FusedLocationProviderClient,
            ctx: MaybeContext,
            factory: MaybeFactory,
            intent: PendingIntent
        ): Maybe<Boolean> {
            return factory.create(
                RemoveLocationIntentUpdatesObservableOnSubscribe(
                    fusedLocationProviderClient,
                    ctx,
                    intent
                )
            )
        }
    }
}